### PR TITLE
Refactor Memory Access Iterators

### PR DESCRIPTION
**Pull request by SimeonEhrig**
- Rename Iterator to Memory Access Strategy
- remove data member from iterator because:
 - not necessary for functionality
 - causes a bug on CPU backends with fancy iterators
 - create unnecessary temporary objects
 - harder to read
 - harder to optimize for the compiler

### CODE REVIEW

**Comment by bussmann** [comment]
I wonder if Vikunja memory access could profit from LLAMA...

**Comment by SimeonEhrig** [comment]
> I wonder if Vikunja memory access could profit from LLAMA...

For this PR, it is not interesting, because the current implementation of the memory access strategies requires 1D memory without padding. But in general, yes. I want to support also multidimensional memory #61 . I think `std::mdspan` is a good idea for the memory access, like @bernhardmgruber already suggested for `llama` and `alpaka`.

**Comment by SimeonEhrig** [comment]
After a discussion with @bernhardmgruber we find out, that `std::mdspan` does not support padding, when the padding is not multiple of the size of a single element. Therefore I will use `alpaka::accessors`, which supports this. 

As side not,  `std::mdspan` supports also other nice features like slicing. But I don't need this feature for vikunja.

**Comment by SimeonEhrig** [comment]
The RP is ready for merging. At the moment, some tests and documentation are missing. This will follow in another PR. This PR needs to be merged, because it is a show stopper for PR #62, #68, #71, #72.

File: include/vikunja/access/BaseStrategy.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        //! \param maximum The first index outside of the iterator memory.
+        ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE BaseStrategy(TIdx const index, TIdx const maximum)
+            : m_index(index)
+            , m_maximum(maximum)
+        {
+        }
+
+        ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE BaseStrategy(const BaseStrategy& other) = default;
+
+        ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE auto operator==(const BaseStrategy& other) const -> bool
```
Comment (j-stephan) [inline include/vikunja/access/BaseStrategy.hpp:40]: @bernhardmgruber recommends to make comparison operators a `friend`:

```suggestion
        ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE friend auto operator==(const BaseStrategy& other) const -> bool
```

Equivalent for the others.

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+     * @tparam TAcc The alpaka accelerator type.
+     * @tparam TIdx The index type
+     *
+     * The memory access policy should provide three values:
+     * - The startIndex of the iterator, which is the first index to use.
+     * - The endIndex of the iterator, which is the last index to use.
+     * - The stepSize of the iterator, which tells how far the iterator should move.
+     */
+    template<typename MemAccessPolicy, typename TAcc, typename TIdx>
+    class PolicyBasedBlockStrategy : public BaseStrategy<TIdx>
```
Comment (j-stephan) [inline include/vikunja/access/PolicyBasedBlockStrategy.hpp:31]: Do you plan to add other block strategies in the future?

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+
+#include "BaseStrategy.hpp"
+
+#include <alpaka/alpaka.hpp>
+
+namespace vikunja::MemAccess
+{
+    /**
+     * A policy based memory access strategy that splits the data access into chunks. Depending on the memory access
+     * policy, these chunks can access the data sequential or in a striding pattern.
```
Comment (j-stephan) [inline include/vikunja/access/PolicyBasedBlockStrategy.hpp:20]:
```suggestion
     * policy, these chunks can access the data sequentially or in a striding pattern.
```

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+     * - The stepSize of the iterator, which tells how far the iterator should move.
+     */
+    template<typename MemAccessPolicy, typename TAcc, typename TIdx>
+    class PolicyBasedBlockStrategy : public BaseStrategy<TIdx>
+    {
+    private:
+        TIdx m_step; /**< The step size of this iterator. */
+    public:
+        /**
+         * Create a policy based block iterator
```
Comment (j-stephan) [inline include/vikunja/access/PolicyBasedBlockStrategy.hpp:37]: Is this still an iterator?

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+         * @param acc The accelerator type to use.
+         * @param problemSize The size of the original iterator.
```
Comment (j-stephan) [inline include/vikunja/access/PolicyBasedBlockStrategy.hpp:39]: Also here.

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        /**
+         * A memory policy for the PolicyBlockBasedIterator that provides grid striding memory access.
+         */
+        struct GridStridingMemAccessPolicy
+        {
+            template<typename TAcc, typename TIdx>
+            ALPAKA_FN_HOST_ACC ALPAKA_FN_INLINE static auto getStartIndex(
+                TAcc const& acc,
+                TIdx const& problemSize __attribute__((unused)),
+                TIdx const& blockSize __attribute__((unused))) -> TIdx const
```
Comment (j-stephan) [inline include/vikunja/access/PolicyBasedBlockStrategy.hpp:164]:
```suggestion
                TIdx const& /* problemSize */,
                TIdx const& /* blockSize */)) -> TIdx const
```

If you really need this consider using `[[maybe_unused]]` instead to make it portable.

File: test/include/vikunja/test/AlpakaSetup.hpp [hunk 7]
```diff
@@ -50,6 +50,62 @@ namespace vikunja
                 , queueAcc{devAcc}
             {
             }
+
+            /**
+             * @brief Allocate 1D memory on host.
```
Comment (j-stephan) [inline test/include/vikunja/test/AlpakaSetup.hpp:55]:
```suggestion
             * @brief Allocate 1D memory on the host.
```

Also in the other places below.

**Review by j-stephan — CHANGES_REQUESTED** [review]
Some suggestions:

File: include/vikunja/access/PolicyBasedBlockStrategy.hpp (hunk 4 above)
Comment (SimeonEhrig) [reply include/vikunja/access/PolicyBasedBlockStrategy.hpp:37]: no

File: include/vikunja/reduce/detail/BlockThreadReduceKernel.hpp [hunk 8]
```diff
@@ -104,11 +104,10 @@ namespace vikunja
 
                     using MemPolicy = TMemAccessPolicy;
                     // Create an iterator with the specified memory access policy that wraps the input iterator.
-                    vikunja::mem::iterator::PolicyBasedBlockIterator<MemPolicy, TAcc, TInputIterator> iter(
-                        source,
-                        acc,
-                        n,
-                        TBlockSize);
+                    using MenIndex = vikunja::MemAccess::BlockStrategy<MemPolicy, TAcc, TIdx>;
```
Comment (j-stephan) [inline include/vikunja/reduce/detail/BlockThreadReduceKernel.hpp:107]: Is `MenIndex` (instead of `MemIndex` intentional)?

Comment (SimeonEhrig) [reply include/vikunja/reduce/detail/BlockThreadReduceKernel.hpp]: You are right. I fixed it.

**Review by j-stephan — APPROVED** [review]

**Pull request merged**